### ISSUE TITLE
Potentially restructure schema

### ISSUE THREAD

**Issue by ericgio** [yekrojam/api#14]
We talked the other day about potentially restructuring the schema a bit, so I wanted to capture that somewhere. A few things discussed:

**Product Hypothesis**
Take a simple approach wrt Orgs, ie: siloing them from each other. Many(?) users will probably only belong to a single org, so a simpler hierarchical model makes things easier on us. This means that each org is basically its own site with no knowledge of other orgs, and a user switching to a different org will just need to go log into that org separately.

**Technical Impact**
- The DB will have separate user documents for each org a person is in.
- We wouldn't need the Memberships collection anymore, since permissions could just live directly in the User schema
- We'd need to update validation so that email was unique on a per-org basis, not globally.
- Depending on how we handle auth, users could still potentially use a single account (Auth0, Slack, FB, etc.) to login. We would just then use those auth credentials to match the user to the correct account.

Any additional thoughts?

**Comment by makinde** [comment]
I'm still going back and forth in my mind. Perhaps the question is what the user thinks they are doing. Are they logging into a site for some org (let's say designer fund) or are they logging into majorkey2.com. The answer to that probably depends tons on what the interface looks like (and maybe what the URL's look like).

If someone goes to designerfund.majorkey2.com, and that looks very different than devcolor.majorkey2.com (or maybe it's even just people.designerfund.com and key.devcolor.org)...then it prob makes sense to do the separate accounts thing. It is presumably modeled the to user as just two different website entirely.

If on the other hand it's majorkey2.com/designedfun and majorkey2.com/devcolor, and they look the same other than the logo at the top, then the user will likely assume that they are logging into the overall site. And that the different org's are merely sub-sites. If it looks like this, then it may be confusing to be logged in on mk2.com/df, but then logged out on mk2.com/color. Additionally, if you go to mk2.com/welcome (perhaps some landing page), are you logged into anything?

In a way, this almost comes back to, does each org (a) have it's own custom built front end, likely hosted on their own domain, or (b) have just a path on one central domain. I guess it could also be (c) some of both. Where there's a central site and 1 off sites. 

Ideally supporting both on the backend would be great. I think the structure with the membership table supports both (while requiring slightly more code to get a list of users for an org). For example:
No membership table:  `GET` `/users?query={org:'...'}`
Current membership table: `GET` `/memberships?query={org:'...'}&populate=user`

The one thing from the technical impact list that would be critical is what to do about emails addresses needing to be unique.

Thoughts?

**Comment by makinde** [comment]
So many edge cases. Here's my latest thought:
- Keep current membership table
- Users can switch between orgs they are a member of in the top nav (via a dropdown)
- Users verify their email address with auth0, then we log them into that email address.
- URL's are something like mk2.com/designerfund or mk2.com/devcolor
- A standalone site for designer fund (or any org) can be built with standalone logins. Those users will auth through some mechanism, and will just be tied to a single membership in that one org.

Advantages
- Users don't have to remember their subdomain. They can login to mk2.com and we can show them their org's.
- Users can see the orgs they are a part of
- Once you are logged in to mk2.com, you are logged in anywhere you go on the site.
- Subdomain certs might be harder to manage
- If we implement it this way, we can always fall back to the simpler model of 1 account per user, per org. But if we start with tying a single user to a single org, we have to do a bunch of work to go the other way.

Disadvantages
- It could be confusing if I have a personal account, and then I want a work account. I'd have to logout and create a new account (with my work email) and know that a couple of the orgs I'm a part of belong to one account, and some to another.
- It's longer to write the query to get the users for an org. Yeah. The upside is that this same structure could be useful for storing members of a group (like a squad) or RSVP's to an event.

There's a brief summary of the different models here: https://stormpath.com/blog/multi-tenancy-for-saas-applications

On a personal note, slack's model had me confused for a long time. If I ever forget the subdomain of a community, I'm SOL. I was additionally confused when I couldn't use the same login to get into all of the workspaces, and i basically use the same username and password for all of them.

**Comment by ericgio** [comment]
> In a way, this almost comes back to, does each org (a) have it's own custom built front end, likely hosted on their own domain, or (b) have just a path on one central domain. I guess it could also be (c) some of both. Where there's a central site and 1 off sites.

This seems to me like the key question. If I had to guess, I think the orgs we've talked to so far would want a), a white-label solution that appears to users as though it's from their own domain, (eg: designerfund.com/directory or directory.designerfund.com). It's hard to say without doing more product research, though.

I'm not sure how best to make a decision and move forward on this without the benefit of more information. In the short-term I'd lean towards going with the simpler solution product- and engineering-wise; I think that would mean not supporting multi-tenancy, but let me know if you disagree.

**Comment by makinde** [comment]
Yeah, I think Designer Fund and Authentic Ventures would likely want their own sites. I also think there are a lot of cases for someone just wanting to turn something on for a community (ERG's, and folks that want to start their own little /dev/color like chapters).

If we go (c), then the question is, can the API support both now. I think it can. The key is probably how authentication is done by the clients. Here's how it could work:
- Auth is done by 3rd party (auth0, let's say) and they provide us with some long unique identifier. That is stored in the user object. When you login through the 3rd party, we/they create a `jwt` that holds that unique identifier, and the API always looks up to see who's the acting user on an api request. If all clients of the API do that, I think it's all good and both scenarios should transparently work on the API.

For building one off sites, each site would just hard code that into the site (or a `.env` variable) so the API can know which org to look at.

Would that unblock things? I imagine you have some questions about what the queries would look like to do various things. If you mention some examples, I can write out what they would look like. I think with the memberships table they'll still be pretty straight forward and possible to get all the data you need in one request.

**Comment by ericgio** [comment]
> Auth is done by 3rd party (auth0, let's say) and they provide us with some long unique identifier. That is stored in the user object. When you login through the 3rd party, we/they create a jwt that holds that unique identifier, and the API always looks up to see who's the acting user on an api request. If all clients of the API do that, I think it's all good and both scenarios should transparently work on the API.

I think auth on login and per-request auth are separate things. The former allows us to know who a user is while the latter allows us to better control where API requests come from. They may intersect since some requests will need to ensure that a given user has the right privileges to complete the request they're making. Mainly noting this because I don't think they need to be coupled in terms of development.

> Would that unblock things?

You mean as far as the login piece? I'm not sure. I feel like you had some questions about my PR for that, so it may be worth revisiting and discussing. I think the main question I have right now is: What are we building, exactly? :) I've just been throwing up some really basic pages and interactions as proofs-of-concept, but haven't thought too deeply about what an MVP would require.

> I imagine you have some questions about what the queries would look like to do various things

Off the top of my head, the main thing I'm not sure about is how to do joins in a request the way the API is currently structured. Lmk if that's in the docs though, and I can just RTFM a bit more carefully.

**Comment by makinde** [comment]
May be easier to chat about the auth stuff live. I think the key thing that the API will care about from the user auth flow is how the requesting user's identity will be conveyed to the API. Will it be some authID? will it be the email address? Will it be the `_id` for the specific user? Each of these would require slightly different code in the API. And the way i've built it, every request to the API will need to check if the entity making the request has permission to see the given object.

> What are we building exactly.

Indeed. We haven't talked about this in a while. Building out the API is pretty useful for me right now. I feel like I'm (a) learning about what it takes to build an API with the current approach (mongoose, express, and other libs) and (b) laying a foundation that could have a few products built on top of it.

For what the specific product is, I've got some ideas and things I'd like to see. That said, I'm still pretty focused on trying/learning new tech right now. Along the way, I'm hopeful that the API I'm working on can be complimentary to the things you're building. For instance, I could imagine you fully taking on a designer fund directory, building it with the API code I've written, while I tinker with a multi-tenant site.

> Joins

The joins are pretty simple. You'll just want to use the `?populate` query param. It's actually just like using a mongoose model directly. You can go down arbitrary layers and filter long the way (just like when doing a mongoose query. Check out [this](https://florianholzapfel.github.io/express-restify-mongoose/#populate) for a rough overview of the REST API and [this](https://mongoosejs.com/docs/populate.html) page describing population in general.

**Comment by ericgio** [comment]
Yeah let's discuss the auth stuff in person, definitely easier. One quick thought:

> Will it be some authID? will it be the email address? Will it be the _id for the specific user?

My PR currently checks the DB for an auth id and creates a user if it doesn't exist. Thinking more about tit, however, I think it probably makes sense to check for an email address since that will be unique. I don't think we can check the `_id` since a 3rd-party auth mechanism won't know about it.

As far as product direction, I'm happy to discuss an actual roadmap if we want to move the product forward. Alternately, I'm also fine with letting your work on the API influence things for now and I can just work on client features that test what you're building to potentially uncover holes.

**Comment by ericgio** [comment]
Sort of related to the above and the concept of singup/login: How do we allow orgs to control who creates an account? The orgs we've been talking to so far would most likely be closed, and wouldn't want just anyone going to a signup page to create an account. Would there be some kind of invite or shadow account system?

**Comment by makinde** [comment]
Yeah, I've been noodling on that question (who can create accounts) for some time. And I think it's (at least from an implementation standpoint) pretty tied in with how auth is done. it doesn't seem like it should be, but I think it is because if someone can create an account for you, we need a way to know which account that is when you actually do login.

**Comment by makinde** [comment]
Looking forward to talking about this tomorrow. I think I have a pretty simple solution to simplify this.

[preview] Let's say you're building a one off site for DesignerFund, and I have another codebase for the multi-tenant version of the service. They can each run their own API servers pointing to their own databases. That actually should simplify almost all the questions we have. Emails can be unique in each database. And since the api exports an express app by default, it's easy to serve the API alongside any given web app (so CORS and cross domain issues go away as well)
